### PR TITLE
[MRG + 1] stac-edit

### PR DESCRIPTION
**Pull request by kowey**
A raft of changes I made while doing another round of annotations cleanup, mainly to 

* make the UI a bit friendlier (target audience: Eric, though): stac-util now split into util (for looking at things) and edit (for modifying the corpus)
* stac-edit now overwrites by default (I've gotten used enough to these tools to trust them a bit more)
* some new utilities to help with new requests: delete-anno (for those redundant rels), split-dialogue (for a complicated move in s1-league1-game3, see svn), and fancier stac-edit move

### CODE REVIEW

File: educe/stac/edit/cmd/delete_anno.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+
+def _delete_in_doc(del_id, doc):
+    """Delete the annotations with the given id in the given document
+
+    NB: modifies doc
+    """
+    pretty_id = anno_id_from_tuple(del_id)
+    is_ok = lambda x: not _is_match(del_id)(x)
+    matches = [x for x in doc.annotations() if not is_ok(x)]
```
Comment (moreymat) [inline educe/stac/edit/cmd/delete_anno.py]: I have tried to unroll the last 2 lines, plus the _is_match function, and arrived at the following formulation which, in my opinion, is really much much easier to understand and maintain:
```python
matches = [x for x in doc.annotations()
                  if anno_id_to_tuple(x.local_id()) == del_id]
```
Explanations could just happen in a comment line.

I have to say I don't feel really comfortable to try and favor certain coding styles, so won't feel really offended if you ignored this comment.

Comment (kowey) [inline educe/stac/edit/cmd/delete_anno.py]: Hah, most of my coding style is accidental.  I thought the point of is_ok was that I use it in several places?

Comment (moreymat) [inline educe/stac/edit/cmd/delete_anno.py]: What about
```python
is_ok = lambda x: anno_id_to_tuple(x.local_id()) != wanted
```
then?

Comment (kowey) [inline educe/stac/edit/cmd/delete_anno.py]: Yup, that's a lot better.

So you can imagine how this came about, right?
There used to be a filter function I was trying to use.
But these days am trying to go with the flow more.
Just didn't notice I could clean out is_match along the way


On 21 May 2015 at 11:59, Mathieu Morey <notifications@github.com> wrote:

> In educe/stac/edit/cmd/delete_anno.py
> <https://github.com/irit-melodi/educe/pull/49#discussion_r30788064>:
>
> > +    an annotation id matches
> > +    """
> > +    def pred(anno):
> > +        "curried second arg"
> > +        return anno_id_to_tuple(anno.local_id()) == wanted
> > +    return pred
> > +
> > +
> > +def _delete_in_doc(del_id, doc):
> > +    """Delete the annotations with the given id in the given document
> > +
> > +    NB: modifies doc
> > +    """
> > +    pretty_id = anno_id_from_tuple(del_id)
> > +    is_ok = lambda x: not _is_match(del_id)(x)
> > +    matches = [x for x in doc.annotations() if not is_ok(x)]
>
> What about
>
> is_ok = lambda x: anno_id_to_tuple(x.local_id()) != wanted
>
> then?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/irit-melodi/educe/pull/49/files#r30788064>.
>



-- 
Eric Kow <http://erickow.com>

File: educe/stac/edit/cmd/rename.py [hunk 2]
```diff
@@ -8,13 +8,13 @@
...
     anno_id
-from ..doc import compute_renames, evil_set_id
-from ..glozz import anno_id_from_tuple, anno_id_to_tuple
-from ..output import save_document
+from educe.stac.util.doc import compute_renames, evil_set_id
+from educe.stac.util.glozz import anno_id_from_tuple, anno_id_to_tuple
+from educe.stac.util.output import save_document
 
 
 def _is_match(wanted):
```
Comment (moreymat) [inline educe/stac/edit/cmd/rename.py]: If I am not mistaken, there should be no more call for this function after your latest commit :-)

File: educe/stac/edit/cmd/nudge_dialogue.py [hunk 3]
```diff
@@ -114,14 +114,14 @@ def _nudge_dialogue(doc, tid, direction):
     """
     prev_turn, turn, next_turn = \
         _window1(lambda x: st.turn_id(x) == tid,
```
Comment (moreymat) [inline educe/stac/edit/cmd/nudge_dialogue.py]: Looking at `_window1` just above, this function could use a small fix of the docstring (match is the item matching the predicate) and a small fix in the code (`before = [None] + now[:-1] ̀).
I cannot comment on lines outside the "diff zone".

File: educe/stac/util/args.py [hunk 4]
```diff
@@ -78,17 +86,17 @@ def read_corpus_with_unannotated(args, verbose=True):
...
+    """Return the output dir specified or inferred from command
+    line args.
 
     We try the following in order:
 
     1. If `--output` is given explicitly, we'll just use/create that
-    2. If the subcommand supports `--overwrite`, and the user specifies it
-       on the command line, the output directory may well be the original
-       corpus dir (*gulp*! Better use version control!)
+    2. If `default_overwrite` is True, or the the subcommand supports
```
Comment (moreymat) [inline educe/stac/util/args.py]: typo

**Comment by moreymat** [comment]
My previous comments aside, LGTM.

File: educe/stac/edit/cmd/rename.py (hunk 2 above)
Comment (kowey) [inline educe/stac/edit/cmd/rename.py]: :-) fixing

**Comment by kowey** [comment]
Thanks for the review!  Always nice to have stuff spotted :-) Tacked on three more patches to hopefully resolve the issues.

**Comment by moreymat** [comment]
[trying to follow sklearn's tested development practice, episode 81983]

**Comment by kowey** [comment]
What happens after MRG +1?

**Comment by moreymat** [comment]
In those big projects, someone has to approve too and change to MRG + 2
then someone merges. Here you can merge when you want.

**Comment by kowey** [comment]
Oh, I see, so this signal works for in-team pull requests too, ie. if the author is in the core team, they can use that as the go-ahead.  And then scales to the team/project growth.  Thanks as ever for clue'ing me in to all these practices.  Eyes open is good.

**Pull request merged**